### PR TITLE
Fix pickle handling for SingletonGate class

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit fixes an oversight in #10314 where the handling of pickle wasn't done correctly. Because the SingletonGate class defines __new__ and based on the parameters to the gate.__class__() call determines whether we get a new mutable copy or a shared singleton immutable instance we need special handling in pickle. By default pickle will call __new__() without any arguments and then rely on __setstate__ to update the state in the new object. This works fine if the original instance was a singleton but in the case of mutable copies this will create a singleton object instead of a mutable copy. To fix this a __reduce__ method is added to ensure arguments get passed to __new__ forcing a mutable object to be created in deserialization. Then a __setstate__ method is defined to correctly update the mutable object post creation.

### CODE REVIEW

File: qiskit/circuit/singleton_gate.py
Comment (jakelishman) [inline qiskit/circuit/singleton_gate.py]: If the `__setstate__` part of the `__reduce__` tuple is `None`, then `__setstate__` doesn't get called at all.  I think the check has no real meaning here.

File: qiskit/circuit/singleton_gate.py
Comment (jakelishman) [inline qiskit/circuit/singleton_gate.py]: Did you have a particular reason to prefer `__setstate__`+`__reduce__` over `__getstate__`+`__setstate__`+`__getnewargs_ex__`?  The current form works fine so doesn't need changing, just asking because the docs suggest we should prefer the higher-level interface.

File: test/python/circuit/test_singleton_gate.py
Comment (jakelishman) [inline test/python/circuit/test_singleton_gate.py:263]: Totally unimportant, but pickle allows serialisation to `bytes` directly, so you can do this as a one-liner if you prefer: `copied = pickle.loads(pickle.dumps(gate))`.

File: test/python/circuit/test_singleton_gate.py
Comment (jakelishman) [inline test/python/circuit/test_singleton_gate.py:264]: I think with this test, we potentially might want to assert a step further and say `gate is copied`?  If we're pickling within the same process that should hold - obviously not so in IPC, but `is` comparisons and the memory reduction aren't relevant over that.

**Review by jakelishman — COMMENTED** [review]
This looks sensible to me, thanks.  The only real comment is the last one about potentially strengthening what we want to assert about the within-process pickle roundtrip.

File: qiskit/circuit/singleton_gate.py
Comment (wshanks) [reply qiskit/circuit/singleton_gate.py]: Actually, I wasn't that familiar with the details of pickling, so I read the docs closely and I think there is a problem with the current form.

When this 3-tuple form is used for reduce, the second element of the tuple is unpacked and passed as the args to the first tuple element. Then `__setstate__` is called on the object returned by that call with the third tuple element passed in. The issue is that it is not guaranteed that all subclasses of `SingletonGate` take `(label, _condition, duration, unit)` as their first four arguments. For example, `test_mutable_pickle` uses `SXGate` which takes `(label, duration, unit, _condition)` (note the order is not the same as `SingletonGate`). When `pickle.load()` is called in the test, the second element of the reduce tuple gets passed to `SXGate` and triggers the mutable creation in `__new__` but then also gets passed to `SXGate.__init__` and initially creates an object with `_condition` and `label` exchanged.

`__getnewargs_ex__` could work if it were guaranteed that all subclasses would accept `(label, _condition, duration, unit)` as keyword arguments. For something more flexible, maybe `__reduce__` could return something like `(object.__new__, (self.__class__,), (self.label, self._condition, self.duration, self.unit))` so that unpickling still bypasses `__init__` but creates a copy it can set the state on?

File: qiskit/circuit/singleton_gate.py
Comment (wshanks) [reply qiskit/circuit/singleton_gate.py]: The docs mention something like this for `__getstate__` [here](https://docs.python.org/3/library/pickle.html#object.__setstate__):

> Note: If [__getstate__()](https://docs.python.org/3/library/pickle.html#object.__getstate__) returns a false value, the [__setstate__()](https://docs.python.org/3/library/pickle.html#object.__setstate__) method will not be called upon unpickling.

but I didn't see it specifically called out for `__reduce__`. I confirmed that `__setstate__` was not called for `test_immutable_pickle`.

File: qiskit/circuit/singleton_gate.py
Comment (jakelishman) [reply qiskit/circuit/singleton_gate.py]: Oh, that's true - I mistook that the first argument is the class object not specifically its `__new__`.  I think the reduction here would be fixed just by making the callable fetch `self.__class__.__new__` exactly to bypass `__init__`.  In practice, I don't think the current form will actually exhibit an issue, because the subsequent call to `__setstate__` will overwrite any mistakes during the `__init__`, but if the `__init__` started doing more type checking on its inputs, it could cause spurious exceptions before that stage.

I think the `__getstate__`/`__setstate__`/`__getnewargs_ex__` form might be a little easier - I think it's just the higher-level formalism of the system I described above with deliberately fetching `__class__.__new__`.

**Comment by wshanks** [comment]
Here is some context for this issue. It was first noticed when the daily run of qiskit-experiments' tests with qiskit's main branch encountered test failures for circuits with conditioned gates (see for example [this run](https://github.com/Qiskit-Extensions/qiskit-experiments/actions/runs/6243811511/job/16949781007)). The tests only failed on Linux because on Windows and macOS the transpiler parallelism is disabled. When parallelism is used, objects are passed between processes using pickle, triggering this issue.

The issue is a little bit worse than mutable gates becoming instances of the immutable singleton when unpickled. The default unpickling procedure given in the documentation is:

```python
def restore(cls, attributes):
    obj = cls.__new__(cls)
    obj.__dict__.update(attributes)
    return obj
```

Every `cls.__new__(cls)` will generate the singleton gate instance, but note that the `obj.__dict__.update(attributes)` will actually mutate the immutable instance, so you could end up with an unexpected condition on all the default `SXGate`s for example.

**Comment by jakelishman** [comment]
Defining `__reduce__` or `__setstate__` prevents that default `restore` from taking effect, I'm pretty sure - that's kind of the point of the methods.

File: test/python/circuit/test_singleton_gate.py
Comment (wshanks) [inline test/python/circuit/test_singleton_gate.py:267]: One thing that surprised me with this issue is that it was not caught in this repo before the singleton gate PR was merged. I expect there is a good set of tests for transpiling conditional gates, but likely the tests all transpile a single circuit and don't hit the pickling issues that come up with parallel transpilation. I wonder if the tests added here are enough or if there should be more tests of pickling and unpickling circuits.

**Comment by wshanks** [comment]
Right, I was just describing how things work now where there is no `__setstate__` or `__reduce__`.

File: test/python/circuit/test_singleton_gate.py
Comment (jakelishman) [reply test/python/circuit/test_singleton_gate.py:267]: There's actually not a _vast_ amount of testing of conditional gates (with `.condition` set), because they typically weren't executable on anything but simulators until IBM hardware grew support for midcircuit measurements and more generalised control flow, at which point we began pivoting to `IfElseOp`.

The bug getting through points to a deficiency in integration testing around conditional gates for sure, and we can add more tests along those lines until `Instruction.condition` is completely removed in 1.0.  Conditionals after transpilation are also harder to write good tests about, because we can't just test for unitary equivalence or anything like that when they're present, so it's easier for stuff to slip through.

File: qiskit/circuit/singleton_gate.py
Comment (wshanks) [reply qiskit/circuit/singleton_gate.py]: I played with `__reduce__` returning `(object.__new__, (self.__class__,), (self.label, self._condition, self.duration, self.unit))`. It kind of worked, but it helped me understand the point about pickle not calling `__init__`. `test_mutable_pickle` created a mutable `SXGate` and set the four attributes appropriately with `__setstate__`. However, the test then failed because the instance did not have a `_name` attribute because only those four attributes were set as `super().__init__()` was never called to set the other attributes. The current was passing the test because returning `self.__class__` from `__reduce__` leads to `__init__` getting called.

So now I understand why you advocate for the higher level methods. I think what would be perhaps most robust is defining `__getnewargs_ex__` to return a special keyword argument like `_create_mutable=True` (which hopefully no subclass would use) and then updating `__new__` to create a copy when this is passed (as well as the other conditions). Then we should not need `__getstate__` or `__setstate__` since the default implementation of copying attributes should be sufficient.

Comment (mtreinish) [reply qiskit/circuit/singleton_gate.py]: I did it this way mostly because my knowledge of the pickle api is dated and I only knew how to do this with `__reduce__`. I wasn't aware of `__getnewargs_ex__`, but that interface is basically exactly what is needed here. I can refactor this to use the newer api.

File: test/python/circuit/test_singleton_gate.py
Comment (mtreinish) [reply test/python/circuit/test_singleton_gate.py:263]: I just don't like using `pickle.dumps` because `s` implies string to me and it's not a string :D

Comment (wshanks) [reply test/python/circuit/test_singleton_gate.py:263]: Rest in peace, Python 2 :slightly_smiling_face:

Comment (jakelishman) [reply test/python/circuit/test_singleton_gate.py:263]: maybe it's an initial from a language that reads right-to-left: "DUMP byteS" instead of "DUMP String" lol

Comment (wshanks) [reply test/python/circuit/test_singleton_gate.py:263]: Here's the Python 2 docstring for dumps :slightly_smiling_face: : "Return the pickled representation of the object as a string, instead of writing it to a file."

File: qiskit/circuit/singleton_gate.py
Comment (mtreinish) [reply qiskit/circuit/singleton_gate.py]: I updated this in: https://github.com/Qiskit/qiskit/pull/10871/commits/3ad38b4ae95f23bae69d47942162345ed71ca6dd we only need to define: `__getnewargs_ex__`  with that defined pickle will call `__new__` correctly so we either correctly use the singleton or the mutable copy returned. After that the default pickle mechanisms work fine as is.

File: qiskit/circuit/singleton_gate.py
Comment (mtreinish) [reply qiskit/circuit/singleton_gate.py]: This isn't relevant anymore with: https://github.com/Qiskit/qiskit/commit/3ad38b4ae95f23bae69d47942162345ed71ca6dd

File: test/python/circuit/test_singleton_gate.py
Comment (mtreinish) [reply test/python/circuit/test_singleton_gate.py:264]: Updated in: https://github.com/Qiskit/qiskit/pull/10871/commits/c24edc1867b68a0c08c0ea5be3561e359387e3d3

**Review by jakelishman — APPROVED** [review]
As far as I'm concerned this seems good to go, but I don't know if Will would like to push for more integration tests like he mentioned.

**Review by wshanks — APPROVED** [review]
I think this is okay as it is. I will open an issue about possible testing. I think the tests added here cover the current cases of concern. What I worry about is someone adding a new `SingletonGate` subclass with an extra slot/attribute not captured in `__getnewargs_ex__`. It might appear fine in single process tests but then misbehave in user code once it starts getting run through pickle for multiprocess transpiling.

**Pull request merged**
